bsp: Added support for IDF v4.4 into ESP-BOX

espzav opened the pull request:
# ESP-BSP Pull Request checklist

- [x] Version of modified component bumped
- [x] CI passing

# Change description
- Added support for IDF v4.4 into ESP-BOX
- Fixed display_audio_photo example (recording)
- Updated Pull Request template for new BSPs
- Used new API for button handling
- New API for display init with configuration
- Removed minimal IDF version from examples

Closes https://github.com/espressif/esp-bsp/issues/151

VojtechBartoska commented:
Hello, @leeebo & @espressif2022, can you please help with review? Thanks

espzav commented:
I changed to draft now, because I will changing the buttons API to new one and I should change CI for test examples for IDF4.4 too.

leeebo commented on examples/display_audio_photo/main/idf_component.yml:
```diff
@@ -1,6 +1,6 @@
+  idf: ">=4.4"
   esp_jpeg: "*"
   esp-box:
     version: "*"
```
To make sure the component manager resolve a correct version:

```suggestion
    version: ">=2.5.0"
```

leeebo commented:
LGTM! Just leave some comments from my point of view

wujiangang commented:
@donghengqaz Please check whether this PR meets your requirements in https://github.com/espressif/esp-bsp/issues/151.

espzav commented:
@leeebo Thank you for the fast review. I fixed all issues. The license should be right, we are using it in all files in BSP.

I reworked buttons for new API too and I am working on changes in CI for testing only some examples in IDF4.4

espzav commented on esp-box/esp-box_idf4.c:
```diff
@@ -0,0 +1,67 @@
+/*
+ * SPDX-FileCopyrightText: 2023 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: CC0-1.0
```
We are using same license in all BSP files.

igrr replied:
@espzav In that case, we should change it to Apache 2.0 license, since you have an apache license in esp-box component: https://github.com/espressif/esp-bsp/blob/c6ae6ebf02bbbadaf583503cea72a7912c2b3bb5/esp-box/LICENSE

espzav replied:
@igrr Thanks for clarify. I see, that we have it same in all BSPs. We should consider it. I will create ticket and I will make it. But problem was, I don't understand licesnes too much. I didn't know, which header should be there. 
Thank you for help.

espzav replied:
Fixed in #195

donghengqaz commented:
API `esp_lcd_panel_disp_on_off` is not included in esp-idf v4.4.3, I see in other code, using `#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 0, 0)` to chose which API is compiled.

donghengqaz commented:
I test in esp-idf v4.4.3, but find it may block in lvgl code `lvgl/src/core/lv_refr.c` boss in single and double buffer mode.

```
if(disp_refr->driver->wait_cb) disp_refr->driver->wait_cb(disp_refr->driver);
```

espzav commented:
@donghengqaz Thank you for your inputs. Do you really need support of IDF 4.4.3 ? We want to test only latest release of IDF4.4

donghengqaz commented:
> @donghengqaz Thank you for your inputs. Do you really need support of IDF 4.4.3 ? We want to test only latest release of IDF4.4

No, I planed to update application to support IDF v4.4.5, but now it is v4.4.3. And you had better set the IDF version requirement to be 4.4.5.

espzav commented:
> > @donghengqaz Thank you for your inputs. Do you really need support of IDF 4.4.3 ? We want to test only latest release of IDF4.4
> 
> No, I planed to update application to support IDF v4.4.5, but now it is v4.4.3. And you had better set the IDF version requirement to be 4.4.5.

Ok, we will make it for IDF 4.4.5

donghengqaz commented:
> Besides this, we need option in Kconfig to configure LVGL task's stack size.

@espzav Please add Kconfig to configure LVGL task's stack size, and I will switch your new version if it is done.

espzav commented:
> > Besides this, we need option in Kconfig to configure LVGL task's stack size.
> 
> @espzav Please add Kconfig to configure LVGL task's stack size, and I will switch your new version if it is done.

I added new function `lv_disp_t *bsp_display_start_with_config(const bsp_display_cfg_t *cfg)` int BSP, where you can put `lvgl_port_cfg_t` with `task_priority`, `task_stack`, `task_affinity`, `task_max_sleep_ms` and `timer_period_ms`. You can use this function instead of `bsp_display_start()`.

igrr commented on examples/display_audio_photo/main/idf_component.yml:
```diff
@@ -1,7 +1,7 @@
 description: BSP Display Audio Photo Example
 dependencies:
-  idf: ">=5.0"
+  idf: ">=4.4"
```
I think you can completely remove `idf: ` line from `main/idf_component.yml`, since the example itself doesn't depend on any particular IDF version; It's enough to specify the dependency in the BSP component

donghengqaz commented:
> > > Besides this, we need option in Kconfig to configure LVGL task's stack size.
> > 
> > 
> > @espzav Please add Kconfig to configure LVGL task's stack size, and I will switch your new version if it is done.
> 
> @donghengqaz It's your PR, mater group has the same requirement ? #190

I don't know more about this, I create an [issue](https://github.com/espressif/esp-bsp/issues/151) long long ago, besides this I find my blocking issue is in SPI bus lock function, not related to BSP.

espzav commented:
Please all, don't forget to approve, when the PR is OK. I cannot merge it till I get at least one approve ;-)

leeebo reviewed (approved):
LGTM!

igrr commented on .github/PULL_REQUEST_TEMPLATE/pr_template_bsp.md:
```diff
@@ -7,6 +7,8 @@
 - [ ] Component was added to CI [upload job](https://github.com/espressif/esp-bsp/blob/master/.github/workflows/upload_component.yml#L17)
 - [ ] New files were added to CI build job
 - [ ] New BSP definitions added to [bsp_ext.py](../examples/bsp_ext.py)
+- [ ] BSP was added to SquareLine
```
```suggestion
- [ ] BSP was [added to SquareLine](https://github.com/espressif/esp-bsp/tree/master/SquareLine/common)
```

igrr commented on .github/PULL_REQUEST_TEMPLATE/pr_template_bsp.md:
```diff
@@ -7,6 +7,8 @@
 - [ ] New files were added to CI build job
 - [ ] New BSP definitions added to [bsp_ext.py](../examples/bsp_ext.py)
+- [ ] BSP was added to SquareLine
+- [ ] BSP is supported with IDF v4.4 and latest
```
```suggestion
- [ ] New BSP supports IDF v4.4.x and later releases
```

igrr commented on esp-box/idf_component.yml:
```diff
@@ -1,13 +1,13 @@
 
 dependencies:
-  idf: ">=5.0"
+  idf: ">=4.4"
```
According to https://github.com/espressif/esp-bsp/pull/192#issuecomment-1643322561, you should change this to 4.4.5:
```suggestion
  idf: ">=4.4.5"
```

espzav replied:
Sure. I think, there should be one CI test for exactly IDF 4.4.5 and second one for latest release IDF 4.4 
Are you agree?

igrr replied:
This is not related to CI test, i think... we can keep CI running for release/v4.4, just need to make sure that this component won't be installed if IDF version is <= 4.4.4 — preventing the issue which @donghengqaz ran into.

igrr commented on esp-box/include/bsp/esp-box.h:
```diff
@@ -162,17 +149,19 @@ extern "C" {
- *      - ESP_ERR_INVALID_ARG   Invalid GPIO number
+ *      - Pointer to codec I2S interface handle or NULL when error occured
  */
-esp_err_t bsp_audio_poweramp_enable(const bool enable);
```
Is it a breaking change to remove this function?

espzav replied:
Yes. The whole audio changes to `esp_codec_dev` are breaking changes.

igrr replied:
In that case, why is the major version of the component not incremented? https://semver.org/#summary

igrr reviewed (commented):
Haven't reviewed the source code changes in detail, just a few small comments about the infrastructure.

espzav commented:
@igrr Thank you for review. I left some answers to your comments.

tore-espressif reviewed (approved):
LGTM!
